### ISSUE TITLE
Refactor cache busting for NGINX

### ISSUE THREAD

**Issue by atsmith813** [forem/forem#12265]
**Is your feature request related to a problem? Please describe.**
A while back when we added NGINX caching, we made note of a [TODO](https://github.com/forem/forem/blob/b45c578ea18ac5b7a53453876b54902f09991558/app/services/edge_cache/bust/nginx.rb#L16-L38) that read:

```
# TODO: (Vaidehi Joshi) - Right now, we are checking that nginx is	
# available on every purge request/call to this bust service. If we are going	
# to bust multiple paths, we should be able to check that nginx is	
# available just once, and persist it on the class with @provider_available?.	
# Then, we could allow for an array of @paths = [] to be passed in,	
# and on single bust instance could bust multiple paths in order.
```

In other words, the issue is that any time we bust caches through NGINX, we first make a `GET` request to see if NGINX is alive and well. If that request comes back as successful, we then make the [`PURGE` request](https://github.com/forem/forem/blob/b45c578ea18ac5b7a53453876b54902f09991558/app/services/edge_cache/bust/nginx.rb#L9).

The problem is, we bust caches a lot and we're using quite a bit of extra resources to check NGINX on _every_ request. IIRC, we originally implemented this because in the early days NGINX wasn't running on Forem instances and cache purging was failing.

**Describe the solution you'd like**
Ideally, we wouldn't need to make this initial `GET` request to NGINX on every request before we purge the cache. Instead, we could explore being more efficient in our checks. Technically, the risk of NGINX not responding is the same as Fastly being down when we make a purge request and we aren't checking if Fastly is alive before each request the same way.

There are a few options we could consider:
1. Stop checking NGINX before each request altogether. And if the `PURGE` request alone fails, we log and handle that case. If we want to be safe, maybe we put an NGINX check in an initializer to make sure it's running when the app starts.
2. Refactor how we are making the check to NGINX.

_Refactoring how we check NGINX_
Looking at option 2 further...there could be 2 ways to go about that.
1. We could initialize a global constant when the app starts to see which caching service is configured (NGINX vs Fastly) and then update that constant anytime the relevant SiteConfigs are updated (someone changing caching services). Similar to above, this check is done once on each deployment/app start.
2. We turn the cache buster into an instance so we can do something like this (pseudo code)
```ruby
buster = EdgeCache::Buster.new
buster.bust(path1)
buster.bust(path2)
buster.bust(path3)
```
Where EdgeCache::Buster would have logic in `initialize` to call NGINX only once to check if it's alive. All the subsequent calls to `#bust` wouldn't ping NGINX again since we'd still be using the same object/instance.

**Describe alternatives you've considered**
We also considered adding the option for the buster to post multiple paths. Thinking about multiple paths...part of the trick here is most of the cache busting is done inside individual services such as [`EdgeCache::BustArticle`](https://github.com/forem/forem/blob/master/app/services/edge_cache/bust_article.rb). We do make calls directly to `EdgeCache::Bust` in a few places, but not as often as we call `#bust` in these individual buster services. So if we modify the buster to accept multiple paths (i.e. an Array), we'd have to change how we call the buster in each service. I imagine it would look something like [`EdgeCache::BustUser`](https://github.com/forem/forem/blob/master/app/services/edge_cache/bust_user.rb) where we do:
```ruby
paths = [
  "/#{username}",
  "/#{username}?i=i",
  "/#{username}/comments",
  "/#{username}/comments?i=i",
  "/#{username}/comments/?i=i",
  "/live/#{username}",
  "/live/#{username}?i=i",
  "/feed/#{username}",
]

# We could change the following to be:
# bust(paths)

paths.each { |path| bust(path) }
```
We could build an Array of paths and call `bust(paths)`. The challenge with this approach is it would require us to refactor all the cache buster services to leverage the multiple path busting setup. And in some ways, it becomes less performant because we loop over things to build an Array of paths (Strings) and then loop again to bust each one. Granted, looping through an Array of Strings of these sizes is arguably negligible wrt performance...but it is a "redundant" effort. Taking a look at [`EdgeCache::BustArticle`](https://github.com/forem/forem/blob/master/app/services/edge_cache/bust_article.rb) again, it's a little...unwieldy with how we generate all the paths to bust :sweat_smile:. There are all kinds of loops and queries going on to make that work.

**Additional context**
I'm not exactly sure where this would all fit in...but one thing we have to keep in mind is thread-safety. AFAIK, no HTTP library in Ruby supports thread safety by default, we'd have to manage that on our own. The Fastly gem isn't thread-safe either (it uses `Net::HTTP` under the hood - not thread-safe). Admittedly, I'm not sure what that looks like if we move to the multiple path approach, or if it would even come into play.

_Big shoutout to @vaidehijoshi for marking this TODO and calling attention to it 👍_

**Comment by vaidehijoshi** [comment]
@atsmith813 Thank you for writing this up in such detail!! If this issue is prioritized into cycle 1, maybe I'll pick this up and finally close the loop on my `TODO` from many moons ago 😂

**Comment by rhymes** [comment]
I don't know what the Fastly gem does internally but if we can manage to issue 8 (taken from Alex's example) HTTP calls concurrently that'd be a big win. Those calls don't share data, we only need to know if any of those failed.

We could also see if we can employ other concurrency techniques (async, ractors, processes) if Fastly doesn't work well with threads. 

All of that to say that my vote goes to having `bust()` accept an array of paths, regardless if internally it busts sequentially (as we do now) or concurrently.

Now that we have a variable that says "NGINX is there or not there" I don't think we need to ping NGINX at the beginning of each bust operation. If the app is configured for NGINX and NGINX is not phisically there then we'd have bigger problems than cache busting, it'd mean the apps is not actually caching, correct?

**Comment by atsmith813** [comment]
@rhymes

>  If the app is configured for NGINX and NGINX is not physically there then we'd have bigger problems than cache busting, it'd mean the apps is not actually caching, correct?

☝️ Correct, spot on!


I'm hesitant to go the route of:
> having bust() accept an array of paths, regardless if internally it busts sequentially (as we do now) or concurrently.

For the context of this issue, I think the goal is to reduce how many times we're making [the `GET` request as a sort of "health check" to NGINX](https://github.com/forem/forem/blob/636f4b0ae1a4a534f1d4c8f99e4794e888829871/app/services/edge_cache/bust/nginx.rb#L16-L37). And maybe the issue of making the purges more efficient could be its own issue.

In general...I think it would be a lot of work/refactoring and it still falls short of the "ideal". Maybe instead of us going down the road of trying to manage threads, concurrency, etc., we can put that work on the caching providers (i.e. Fastly and NGINX). And we can do that by refactoring cache purges to use keys/surrogate keys* instead of us purging "manually" with HTTP calls.

Technically, we're still operating on borrowed time because we're using Fastly's now the deprecated method of "manual" purging 😅 . The reason we haven't gone down this road already is mostly that NGINX only supports purging by key out of the box only in the paid(enterprise?) version. So we need to build out that functionality ourselves (Lua <> OpenResty) or improve caching on the ForemCloud side in a different way.

---

***Purging by key**
_for those who aren't as familiar with caching and purging, here's a little crash course on what I mean by "purging by key"_.

When we cache content, we have the ability to tag that cache with keys. The keys can be anything, `article-id-1`, `user-1-article-2`, etc. And you can do multiple. To purge by key would be to send **one** request to bust all caches that have the key we specify. So, effectively, we could bust all the caches we need with a single call if we setup tags/keys correctly.

Taking a look at a quick example...when we purge caches for an article once it's updated, there are _many_ paths that we bust. A subset looks like this:
```ruby
bust("#{article.path}/")
bust("#{article.path}?i=i")
bust("#{article.path}/?i=i")
bust("#{article.path}/comments")
bust("#{article.path}?preview=#{article.password}")
bust("#{article.path}?preview=#{article.password}&i=i")
```

As you may be able to see, we start busting all paths related to that article. If we tagged each of these paths with something like `article-id-1` and then issued a single purge request by that key, Fastly would bust all those caches for us - no concurrency or anything 🕺 . The same is true for NGINX, if/when we get that setup.

Anyway, maybe we can consider putting resources into this approach as well to optimize cache purging instead of or in addition to managing concurrency.

**Comment by vaidehijoshi** [comment]
> 2. We turn the cache buster into an instance so we can do something like this (pseudo code)
> 
> ```ruby
> buster = EdgeCache::Buster.new
> buster.bust(path1)
> buster.bust(path2)
> buster.bust(path3)
> ```
> 
> Where EdgeCache::Buster would have logic in `initialize` to call NGINX only once to check if it's alive. All the subsequent calls to `#bust` wouldn't ping NGINX again since we'd still be using the same object/instance.

I've been marinating on this for a few days and the more that I think about it, the more I feel that this is the best approach. It improves the current scenario of "ping NGINX each time even though we don't have to" without doing a major overhaul of how we do cache busting (which, as @atsmith813 explained in the issue description, would be a major undertaking). Seems like we can check if NGINX is awake once per "a set of busts" so to speak.

I also think that it would be:
1) not too time-consuming of an undertaking, and
2) not something we'd need to majorly rework if we were to move to the paid version of OpenResty down the road.

**Comment by rhymes** [comment]
@atsmith813 I agree 100%

> In general...I think it would be a lot of work/refactoring and it still falls short of the "ideal". Maybe instead of us going down the road of trying to manage threads, concurrency, etc., we can put that work on the caching providers (i.e. Fastly and NGINX). And we can do that by refactoring cache purges to use keys/surrogate keys* instead of us purging "manually" with HTTP calls.

❤️ 

If that's possible, yes! As you know I've spent quite a bit of time setting up surrogate keys with Fastly for the DEV API and you've spent even more time working on caching issues and yeah, it should be the upstream server in charge of deleting data concurrently. My suggestion was because I understood we had no choice but to lean in busting paths, but if we can switch to keys then working on dealing with concurrency client side makes way less sense.

> Technically, we're still operating on borrowed time because we're using Fastly's now the deprecated method of "manual" purging 😅 . 

So, if we're working on borrowed time and DEV isn't going to switch from Fastly anytime soon, what would you recommend as the first priority? Refactoring the NGINX code? Refactoring DEV to use surrogate keys everywhere? Switching DEV away from Fastly and keeping URLs?

> The reason we haven't gone down this road already is mostly that NGINX only supports purging by key out of the box only in the paid(enterprise?) version (which isn't friendly to self-hosted/FOSS usage of Forem). So we need to build out that functionality ourselves (Lua <> OpenResty) or improve caching on the ForemCloud side in a different way, which is in progress by systems (exploring the paid version of OpenResty as a layer over all of ForemCloud).

That's the biggest issue I see though. Is it worth it going down that route if NGINX doesn't really support tag based/surrogate key based caching? I see Varnish supports it instead, which we don't use (and I'm not particularly advocating for, it just popped up in the search results): https://book.varnish-software.com/4.0/chapters/Cache_Invalidation.html

TLDR; if this issue isn't that big as @vaidehijoshi suggests in the previous comment, I think we should do it nonetheless. Then we should figure out what's the long term strategy and adjust accordingly because it really depends if we want to go all on with paths or we want to use surrogate keys.

Thank you again @atsmith813!

**Comment by vaidehijoshi** [comment]
I’m not going to assign myself to this just _yet_ because I haven’t started work on this,  and our workflow explicitly stated that you should only assign yourself to tasks that you’re working on currently...but I am going to leave a comment here (in case anyone looks at this issue) to say that I’d like to pick up work on this on Monday :)

**Comment by atsmith813** [comment]
@rhymes well said! It sounds like we're on the same page.

> So, if we're working on borrowed time and DEV isn't going to switch from Fastly anytime soon, what would you recommend as the first priority? Refactoring the NGINX code? Refactoring DEV to use surrogate keys everywhere while it's on Fastly? Switching DEV away from Fastly and keeping URLs?

I believe this refactor and updating the purge methods to be safe and get off of "borrowed time" (https://github.com/forem/forem/issues/12326) makes sense.

As for the other parts, I think waiting to see what we ultimately do on the ForemCloud side for caching will (more or less) push us in a certain direction. Ideally, we would use keys or the same approach for all caching services (Fastly, NGINX, OpenResty paid version, etc). So once we settle on those architecture decisions, the answers to these questions will hopefully be more obvious to us.

> That's the biggest issue I see though. Is it worth it going down that route if NGINX doesn't really support tag based/surrogate key based caching? I see Varnish supports it instead, which we don't use (and I'm not particularly advocating for, it just popped up in the search results): https://book.varnish-software.com/4.0/chapters/Cache_Invalidation.html

Spot on! That is the issue. I think once we know for certain what we're going to do, we can see what's supported and what isn't. For example, I could see a future where we leave the FOSS of NGINX _without_ key purging in place to better support self-hosted instance of Forem, and leverage key purging with OpenResty on ForemCloud.

And we actually explored VCL with systems initially instead of/alongside OpenResty and Lua. IIRC, we didn't go that direction because 1) it was much more involved to get set up and we needed something quick and 2) Fastly uses it's own, custom fork of VCL so even if we used VCL, the synergies with how Fastly does caching wouldn't be present since "regular VCL" is notably different (I believe I recall Fastly implementing custom VCL methods for purging with keys so re-creating that ourselves would've been pretty intense, relatively speaking).

**TLDR**
I think this issue and https://github.com/forem/forem/issues/12326 are both "quick" fixes with notable improvements we can make now. And then, as you mentioned, once we know more about what systems is going to do for caching on ForemCloud:
> we should figure out what's the long term strategy and adjust accordingly because it really depends if we want to go all on with paths or we want to use surrogate keys.

**Comment by vaidehijoshi** [comment]
I was really hoping I'd be able to tackle this, but my plate is honestly just too full with Product leading and other management priorities 😭 I don't want to block anyone from taking this task on, and I apologize if my assignment to it for so long caused any frustrations.

Anyone else on the Forem core team is free to tackle this if they'd like to! Maybe just ping me for review so I can live vicariously through your code 🙃

**Comment by rhymes** [comment]
End of an era :D
